### PR TITLE
8312200: Fix Parse::catch_call_exceptions memory leak

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi,

We used to allocate some `GrowableArray`s onto the `node_arena` in `Parse::catch_call_exceptions`. This leaves the allocated memory until the compilation has finished, potentially increasing the maximum memory usage. I've fixed this by allocating the memory on a temporary `Arena` instead. I also switched the `GrowableArray`s themselves to being stack allocated.

Also a few stylistic issues were addressed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312200](https://bugs.openjdk.org/browse/JDK-8312200): Fix Parse::catch_call_exceptions memory leak (**Bug** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14921/head:pull/14921` \
`$ git checkout pull/14921`

Update a local copy of the PR: \
`$ git checkout pull/14921` \
`$ git pull https://git.openjdk.org/jdk.git pull/14921/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14921`

View PR using the GUI difftool: \
`$ git pr show -t 14921`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14921.diff">https://git.openjdk.org/jdk/pull/14921.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14921#issuecomment-1640057444)

### CODE REVIEW

File: src/hotspot/share/opto/doCall.cpp [hunk 1]
```diff
@@ -790,46 +790,47 @@ void Parse::catch_call_exceptions(ciExceptionHandlerStream& handlers) {
...
+  GrowableArray<int> saw_unloaded(&tmp_mem, 8, 0, -1);
 
   bool default_handler = false;
   for (; !handlers.is_done(); handlers.next()) {
-    ciExceptionHandler* h        = handlers.handler();
-    int                 h_bci    = h->handler_bci();
-    ciInstanceKlass*    h_klass  = h->is_catch_all() ? env()->Throwable_klass() : h->catch_klass();
+    ciExceptionHandler* h       = handlers.handler();
+    int                 h_bci   = h->handler_bci();
+    ciInstanceKlass*    h_klass = h->is_catch_all() ? env()->Throwable_klass() : h->catch_klass();
```
Comment (jdksjolen) [inline src/hotspot/share/opto/doCall.cpp:802]: Reduced the alignment here by 1, I think that `h_extype` used to be put here, which is what caused an unnecessary extra space.

File: src/hotspot/share/opto/doCall.cpp [hunk 2]
```diff
@@ -790,46 +790,47 @@ void Parse::catch_call_exceptions(ciExceptionHandlerStream& handlers) {
...
            so don't duplicate. Duplication will cause the CatchNode to be
            unnecessarily large. See 4713716. */
         continue;
       } else {
-        saw_unloaded->append(h_bci);
+        saw_unloaded.append(h_bci);
       }
     }
-    const Type*         h_extype = TypeOopPtr::make_from_klass(h_klass);
+    const Type* h_extype = TypeOopPtr::make_from_klass(h_klass);
```
Comment (jdksjolen) [inline src/hotspot/share/opto/doCall.cpp:814]: Removed the alignment here, as it's far away from the previously declared variables in this scope.

**Comment by jdksjolen** [comment]
Passes tier1.

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by jdksjolen** [comment]
Thank you.

/integrate

**Pull request closed without merging**